src/dbnode/storage/index.go

```diff
@@ -475,17 +498,213 @@ func (i *nsIndex) Tick(c context.Cancellable) (namespaceIndexTickResult, error)
 	return result, multiErr.FinalError()
 }
 
-func (i *nsIndex) ReplaceBlock(
-	blockStart time.Time,
-	segments []segment.Segment,
+func (i *nsIndex) Flush(
+	flush persist.IndexFlush,
+	shards []databaseShard,
 ) error {
 	i.state.RLock()
-	defer i.state.RUnlock()
 	if !i.isOpenWithRLock() {
+		i.state.RUnlock()
 		return errDbIndexUnableToFlushClosed
 	}
 
-	return fmt.Errorf("not implemented")
+	flushable := make([]index.Block, 0, len(i.state.blocksByTime))
+	for _, block := range i.state.blocksByTime {
+		if !i.canFlushBlock(block, shards) {
+			continue
+		}
+		flushable = append(flushable, block)
+	}
+	i.state.RUnlock()
+
+	var evictResults index.EvictMutableSegmentResults
+	for _, block := range flushable {
+		immutableSegments, err := i.flushBlock(flush, block, shards)
+		if err != nil {
+			return err
+		}
+		// Make a result that covers the entire time ranges for the
+		// block for each shard
+		fulfilled := make(result.ShardTimeRanges, len(shards))
+		for _, shard := range shards {
+			fulfilled[shard.ID()] = xtime.NewRanges(xtime.Range{
+				Start: block.StartTime(),
+				End:   block.EndTime(),
+			})
+		}
+		// Add the results to the block
+		results := result.NewIndexBlock(block.StartTime(), immutableSegments,
+			fulfilled)
+		if err := block.AddResults(results); err != nil {
+			return err
+		}
+		// It's now safe to remove the mutable segments as anything the block
+		// held is covered by the owned shards we just read
+		evictResult, err := block.EvictMutableSegments()
+		evictResults.Add(evictResult)
+		if err != nil {
+			// deliberately choosing to not mark this as an error as we have successfully
+			// flushed any mutable data.
+			i.logger.WithFields(
+				xlog.NewField("err", err.Error()),
+				xlog.NewField("blockStart", block.StartTime()),
+			).Warnf("encountered error while evicting mutable segments for index block.")
```

Comment:
nit: Remove trailing period from log for consistency with other logs?

Reply:
sure thing

Reply:
done